Fixed coverage map issues; better aligned with LLVM APIs

richkadel opened the pull request:
Found some problems with the coverage map encoding when testing with more than one counter per function.

While debugging, I realized some better ways to structure the Rust implementation of the coverage mapping generator. I refactored somewhat, resulting in less code overall, expanded coverage of LLVM Coverage Map capabilities, and much closer alignment with LLVM data structures, APIs, and naming.

This should be easier to follow and easier to maintain.

r? @tmandry 

Rust compiler MCP rust-lang/compiler-team#278
Relevant issue: #34701 - Implement support for LLVMs code coverage instrumentation

Commit:
Fixed coverage map issues; better aligned with LLVM APIs

Found some problems with the coverage map encoding when testing with
more than one counter per function.

While debugging, I realized some better ways to structure the Rust
implementation of the coverage mapping generator. I refactored somewhat,
resulting in less code overall, expanded coverage of LLVM Coverage Map
capabilities, and much closer alignment with LLVM data structures, APIs,
and naming.

This should be easier to follow and easier to maintain.

richkadel commented:
@wesleywiser @tmandry 

I have more to upload, but I'm going to do separate PRs for the recent changes I've made to add more counters based on MIR branching.

richkadel commented:
r? @tmandry

Commit:
Refactor MIR coverage instrumentation

Lays a better foundation for injecting more counters in each function.

richkadel commented:
@tmandry As discussed, I just added a second commit covering the recent refactoring of MIR pass in instrument_coverage.rs.

This PR had no effect on the `mir-opt` or `run-make-fulldeps` test results.

I'll submit a "WIP"/draft PR separately, with the changes we reviewed, injecting counters before each BasicBlock. I have to regenerate the test expected results for that one.

richkadel commented:
Interesting side note... The format check failed, even though I had run `./x.py fmt`!

I had copied a snapshot of the `instrument_coverage.rs` source to my toplevel `rust/` directory. Files not checked into `git` (like that file) are "ignored" by `./x.py fmt`, which is good. _But_ apparently this ignores _all_ files with names matching the ignored file names. Not just the one at the path `./instrument_coverage.rs`.

```
$ ./x.py fmt
Updating only changed submodules
Submodules updated in 0.06 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
skip untracked path instrument_coverage.rs during rustfmt invocations
Build completed successfully in 0:00:09
```

I had to deleting `./instrument_coverage.rs` and _then_ run `./.x.py fmt`.

wesleywiser commented on src/librustc_codegen_llvm/coverageinfo/mapgen.rs:
```diff
@@ -32,174 +23,123 @@ use std::path::PathBuf;
-        Self { cx, filenames: Vec::new(), filename_to_index: FxHashMap::<CString, u32>::default() }
+impl CoverageMapGenerator {
+    fn new() -> Self {
+        Self { filenames: Vec::new(), filename_to_index: FxHashMap::<CString, u32>::default() }
```
Nit: You can just say `FxHashMap::default()` here.

richkadel replied:
done

wesleywiser commented on src/librustc_codegen_llvm/coverageinfo/mod.rs:
```diff
@@ -100,107 +105,150 @@ impl CoverageInfoBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
-    pub raw: &'a mut llvm::coverageinfo::SmallVectorCounterExpression<'a>,
-}
+/// Aligns to C++ struct llvm::coverage::Counter::CounterKind.
+/// The order of discrimiators is important.
```
Nit: these are usually called "variants". We should probably go ahead and give them explicit values as well.

richkadel replied:
Since I'm numbering the "variants" now, it's probably not necessary to say, "The order of variants is important." I think I'll just remove that line, if OK with you.

wesleywiser replied:
Yeah, that makes sense!

richkadel replied:
done

wesleywiser commented on src/librustc_codegen_llvm/coverageinfo/mod.rs:
```diff
@@ -100,107 +105,150 @@ impl CoverageInfoBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
+#[repr(C)]
+enum RegionKind {
+    /// A CodeRegion associates some code with a counter
+    CodeRegion,
```
```suggestion
    CodeRegion = 0,
```

etc

richkadel replied:
@wesleywiser Will do.

Also, what do you think about providing a link to the in-tree LLVM source this is mapping to? For example:

```rust
/// Aligns to C++ struct [llvm::coverage::Counter::CounterKind](https://github.com/rust-lang/llvm-project/blob/rustc/10.0-2020-05-05/llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h#L203-L221)
```

Should result in rustdoc:

> Aligns to C++ struct [llvm::coverage::Counter::CounterKind](https://github.com/rust-lang/llvm-project/blob/rustc/10.0-2020-05-05/llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h#L203-L221)

I think it's particularly convenient to see exactly what the originating C++ definition was.

Originally, I left it out because there is a slight risk someone might assume the link maps to the _current_ Rust snapshot, which could be misleading if, for some reason, one of the Rust-replicated enum or struct definitions diverges, in some meaningful way, in a future version/shapshot.

Rust's fork of LLVM doesn't use `master` apparently (per https://github.com/rust-lang/llvm-project/blob/master/README.md) and there's no `latest` tag or anything like that, from what I see. So I don't think there's a way to make this a version-less link.

richkadel replied:
(Note: I made some small edits to the comment above, fixing errors in the link text and URL start line)

wesleywiser replied:
I think linking to a specific version is fine. IIRC we do similar things in other places so there's precedent for that.

richkadel replied:
done

wesleywiser commented on src/librustc_codegen_ssa/coverageinfo/map.rs:
```diff
@@ -1,289 +1,376 @@
-pub enum CounterOp {
-    // Note the order (and therefore the default values) is important. With the attribute
-    // `#[repr(C)]`, this enum matches the layout of the LLVM enum defined for the nested enum,
-    // `llvm::coverage::CounterExpression::ExprKind`, as shown in the following source snippet:
-    // https://github.com/rust-lang/llvm-project/blob/f208b70fbc4dee78067b3c5bd6cb92aa3ba58a1e/llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h#L146
-    Subtract,
-    Add,
+enum CounterKind {
+    Zero,
+    CounterValueReference,
+    Expression,
```
Same thing here

richkadel replied:
done

wesleywiser commented on src/librustc_codegen_ssa/coverageinfo/map.rs:
```diff
@@ -1,289 +1,376 @@
-            end_byte_pos: 0,
-        }
+/// Aligns to C++ struct llvm::coverage::CounterExpression::ExprKind.
+/// The order of discriminators is important.
+#[derive(Copy, Clone, Debug)]
+#[repr(C)]
+pub enum ExprKind {
+    Subtract,
+    Add,
```
Same here

richkadel replied:
done

wesleywiser commented on src/librustc_codegen_ssa/coverageinfo/map.rs:
```diff
@@ -1,289 +1,376 @@
 }
 
-impl Ord for CoverageLoc {
+impl Ord for Region {
```
You can probably `#[derive(PartialOrd, Ord)]`

richkadel replied:
@wesleywiser I don't think I can, since `Loc` doesn't implement them. (Let me know if I'm missing something.)

richkadel replied:
(Also, note, Loc contains an `Lrc<SourceFile>` and `SourceFile` doesn't implement `Ord` or `Eq` either. I don't feel confident asserting an ordering scheme for a the `Loc` type since choosing how to impose ordering on its `SourceFile` might have unexpected side effects.)

wesleywiser replied:
Ah, ok. Makes sense.

wesleywiser commented on src/librustc_codegen_ssa/coverageinfo/map.rs:
```diff
@@ -1,289 +1,376 @@
 }
 
-impl fmt::Display for CoverageLoc {
+impl PartialEq for Region {
```
Same with `PartialEq` and `Eq`

richkadel replied:
@wesleywiser same problem

wesleywiser reviewed (commented):
Left a few small comments

richkadel commented on src/librustc_codegen_ssa/coverageinfo/map.rs:
```diff
@@ -1,289 +1,376 @@
+// for the example binary in the crates.io crate `json5format`, I had a couple of
+// notable problems:
+//
+//   1. I saw a lot of coverage spans in `llvm-cov show` highlighting regions in
+//      various comments (not corresponding to rustdoc code), indicating a possible
+//      problem with the byte_pos-to-source-map implementation.
```
That's a good question, and something I've considered too, but I suspect it may be only loosely related.

I haven't attempted a deep investigation of this yet. I've been focusing on other priorities.

But related to your question, this is also why I added the comments you reviewed below on line 179, as well. I want to see if I _really_ need to somehow bring in the source from external crates (which seems to go against the current flow of the rustc architecture), or if I can rely on the external crates' own coverage instrumentation (which would be encoded in a coverage map in that library), and possibly avoid re-instrumenting those dependent references, which would avoid the problem of requesting the source of external crates.

At the moment, I don't see how that's possible with parameterized functions, for example. So I think that theory is wrong, but when I'm ready to look into this more closely (_not now_) I may try to pull that thread and see if it leads to a better answer.

The notes are left here for when I'm ready to make this a priority (though I'm constantly reviewing these thoughts when I'm touching code like the span issue).

richkadel commented on src/librustc_codegen_ssa/coverageinfo/map.rs:
```diff
@@ -1,289 +1,376 @@
+//      have here only applies to the local crate, and I know I have coverages that
+//      reference external crates.
+//
+//          I still don't know if I fixed the hash problem correctly. If external crates
+//          implement the function, can't I use the coverage counters already compiled
+//          into those external crates? (Maybe not for generics and/or maybe not for
```
Agreed. See my reply above. I think we're saying some of the same things, particularly, my paragraph starting:

> At the moment, I don't see how that's possible with parameterized functions, for example.

tmandry replied:
I'm saying that the solution we already use for code (weak linkage) might work for the counters, too.

richkadel replied:
Yes, I understand. Sounds good. I'll look into it in a future PR.

richkadel commented on src/librustc_codegen_llvm/coverageinfo/mapgen.rs:
```diff
@@ -32,174 +23,123 @@ use std::path::PathBuf;
+        // `file_id` (indexing files referenced by the current function), and construct the
+        // function-specific `virtual_file_mapping` from `file_id` to its index in the module's
+        // `filenames` array.
+        counter_regions.sort_by_key(|(_counter, region)| *region);
```
Great point. I don't need a stable sort.

richkadel replied:
done

richkadel commented on src/librustc_codegen_llvm/coverageinfo/mod.rs:
```diff
@@ -100,107 +105,150 @@ impl CoverageInfoBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
+/// operands), the source file (an indirect index into a "filenames array", encoded separately),
+/// and source location (start and end positions of the represented code region).
+///
+/// Aligns to C++ struct llvm::coverage::CounterMappingRegion.
```
I'm going to add links to the C++ structs and enums that these Rust `repr(C)` versions map to. AFAIK, there isn't much more we can do to automate checking for changes, but otherwise, having Rust representation of LLVM types is not uncommon (see, for example, src/librustc_codegen_llvm/llvm/ffi.rs).

I suppose one benefit to the version-specific, line-targeted rustdoc links is that, in the future, when someone is upgrading to a new branch of the rust LLVM fork, they can grep the codebase for mentions of the old branch, update them, re-check the C++ definitions, and fix both line references and (if necessary) type layout changes.

richkadel replied:
(See the discussion here: https://github.com/rust-lang/rust/pull/74733#discussion_r461788404)

richkadel replied:
I think the latest changes address your feedback

tmandry replied:
Okay, we may want to consider putting these in the file you mention so as not to make upgrading LLVM harder. Or in a coverageinfo/ffi.rs which makes it easier to find.

richkadel replied:
Added one more separate commit for this. Check it out and let me know if this looks better. Thanks!

richkadel commented on src/librustc_codegen_llvm/coverageinfo/mapgen.rs:
```diff
@@ -32,174 +23,123 @@ use std::path::PathBuf;
+                let filename = CString::new(file_path.to_string_lossy().to_string())
+                    .expect("null error converting filename to C string");
+                debug!("  file_id: {} = '{:?}'", current_file_id, filename);
+                let filenames_index = match self.filename_to_index.get(&filename) {
```
@tmandry I actually started with that, but there are a couple of problems: (1) there are borrow checker errors on `self` that I'd still need to work around in some way, and (2) the key is a `CString` (not a reference), and `entry()` does not take a reference, but `get()` will. Even if I could work around the borrow check errors, I would have to `clone()` the `filename` passed to `entry()` every time. But with the `get()` implementation, I only have to clone filename if it's a new entry. (A small performance optimization but, unless I'm missing something, I don't think there's a functional advantage to using `entry()` vs `get()` here.)

Let me know if I'm missing something.

richkadel commented on src/librustc_codegen_llvm/coverageinfo/mod.rs:
```diff
@@ -100,107 +105,150 @@ impl CoverageInfoBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
+impl CounterMappingRegion {
+    pub fn code_region(
+        counter: Counter,
+        file_id: u32,
```
I thought about that as well, but was concerned the `newtype_index!` resulting type would have a struct layout that might not align correctly when embedded in a `repr(C)` struct.

But the macro's docs say "the index uses a u32", so I'll give it a shot.

richkadel replied:
(Actually, looking at implementation, this issue may not actually be relevant.)

richkadel replied:
done, in the separate commit https://github.com/rust-lang/rust/pull/74733/commits/b58afc088f34341cdffad8d12bf1a13331ff7deb

Commit:
FunctionCoverage: improve type checking with newtype_index types

richkadel commented:
I responded to all but one of the feedback comments (if a change was needed) in https://github.com/rust-lang/rust/pull/74733/commits/20f55c193dea0a893c869dd7a9cf3e2298635221, but I pushed a separate commit https://github.com/rust-lang/rust/pull/74733/commits/b58afc088f34341cdffad8d12bf1a13331ff7deb with changes to use `newtype_index!` types in `FunctionCoverage` per Tyler's suggestion. I think the change seems OK, and probably has some type-checking advantages, but I wanted you guys to have the option of comparing the two.

richkadel commented on src/librustc_codegen_llvm/coverageinfo/mapgen.rs:
```diff
@@ -32,174 +23,123 @@ use std::path::PathBuf;
-    let mut coverage_writer = CoverageMappingWriter::new(cx);
-
     let function_coverage_map = cx.coverage_context().take_function_coverage_map();
+    if function_coverage_map.len() == 0 {
```
done

Commit:
Moved structs/enums with repr(C) to LLVM types into ffi.rs crates

Some were in librustc_codegen_llvm, but others are not tied to LLVM, so
I put them in a new crate: librustc_codegen_ssa/coverageinfo/ffi.rs

tmandry commented on src/librustc_mir/transform/instrument_coverage.rs:
```diff
@@ -171,17 +170,25 @@ impl<'a, 'tcx> Instrumentor<'a, 'tcx> {
-        self.inject_counter(body_span, next_block);
+        let _ignore = mir_body;
+        let id = self.next_counter();
+        let function_source_hash = self.function_source_hash();
```
Binding is_cleanup makes the code more readable, but I'm not sure this does

richkadel replied:
Got it. Thanks for approving anyway. These lines are completely replaced in the follow-up PRs, but your point may still be relevant in those updates, so I'll try to keep it in mind.

tmandry commented:
I think this is ready to merge.
@bors r+

bors commented:
:pushpin: Commit 5b2e2b25e41afbbd0ad803f7986d8559ef649a7e has been approved by `tmandry`

<!-- @bors r=tmandry 5b2e2b25e41afbbd0ad803f7986d8559ef649a7e -->
<!-- homu: {"type":"Approved","sha":"5b2e2b25e41afbbd0ad803f7986d8559ef649a7e","approver":"tmandry"} -->

bors commented:
:hourglass: Testing commit 5b2e2b25e41afbbd0ad803f7986d8559ef649a7e with merge db0492ace429cfeb3567e2c04e300be7df9972ff...
<!-- homu: {"type":"BuildStarted","head_sha":"5b2e2b25e41afbbd0ad803f7986d8559ef649a7e","merge_sha":"db0492ace429cfeb3567e2c04e300be7df9972ff"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/925180945), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35682)
Approved by: tmandry
Pushing db0492ace429cfeb3567e2c04e300be7df9972ff to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"tmandry","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35682","checks-actions":"https://github.com/rust-lang-ci/rust/runs/925180945"},"merge_sha":"db0492ace429cfeb3567e2c04e300be7df9972ff"} -->

richkadel commented:
@wesleywiser @tmandry FYI - In this version of Rust Coverage, I intentionally reintroduced the "assert_eq" that the start and end of a code region are from the same file. I know this breaks when attempting to compile larger crates with dependencies, but to ignore the problem simply produced bad results.

The good news is, I found a way to fix the problem, and I'm getting clean coverage and code regions across both local and external crates now, when I compile a crate with dependencies!

I'm going to push a small PR tomorrow with this fix.

I mentioned this to Tyler tonight as well: I know the fix works better and produces clean results. I'm not 100% sure it is producing _complete_ results yet.

After I post the new PR so you can see what I'm doing, I'll explain the new solution, why I think it works, and what I'm not sure is covered. (I'm sure I can create some tests to confirm things, either way.)

wesleywiser commented:
@richkadel By the way, thanks for splitting this work up into lots of PRs! I'm sure it would probably be easier for you to just iterate on the code and dump it over the fence but doing it this way makes reviewing it much more manageable. 

When you have this a bit closer to working (which sounds like it may be very soon!), we might want to think about adding a lot more tests both to ensure things are working correctly and to make sure they keep working correctly in the future.

richkadel commented:
@wesleywiser On the subject of tests, since testing the end product requires building and then running a binary, and then running llvm tools, plus the `rust-demangler` tool, I _think_ `run-make-fulldeps` is the right place for building more tests.

Generating the `expected` results is manual and a bit awkward.

Do you think it would be reasonable for me to update the test framework to allow `--bless` to be used with `run-make-fulldeps`, causing it to simply set an environment variable (like, `RUST_TEST_BLESS`), and then `Makefile`s could optionally check that env setting, and optionally recreate any needed "expected" results?

I would use this if it existed.

wesleywiser commented:
@richkadel That seems fine to me. I don't think I've ever touched those tests but I think in general, we'd like to see `--bless` or some equivalent be available everywhere.

Full integration tests would be great but I'd also like to see more `mir-opt` tests since they're going to be the most precise.

richkadel commented:
> Full integration tests would be great but I'd also like to see more `mir-opt` tests since they're going to be the most precise.

I completely agree. My first paragraph was poorly worded. I definitely plan to include a lot more mir-opt tests as well.
